Test support for cloud management changes

car-roll opened the pull request:
<!-- Please describe your pull request here. -->
The moving of cloud management to multiple pages (https://github.com/jenkinsci/jenkins/pull/7658) appear to have broken some tests.

Bug also discovered when changing cloud name. See https://issues.jenkins.io/browse/JENKINS-71737

Also upstream of #888 

cc @Vlatombe 

### Testing done

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

car-roll commented on src/test/java/hudson/plugins/ec2/ConfigurationAsCodeTest.java:
```diff
@@ -34,13 +34,14 @@ public class ConfigurationAsCodeTest {
-        assertNotNull(ec2Cloud);
-        assertEquals(0, ec2Cloud.getTemplates().size());
-    }
+    // Jenkins 2.403: Cloud Nodes cannot be created with an empty name
```
Unable to catch the illegal argument exception when the cloud node is created. Wasn't sure if there was a way to do that for this particular test

raul-arabaolaza replied:
Maybe you can check the instance logs and search for the exception there?. More important, is this change in behaviour expected or a side effect?

car-roll replied:
But isn't the point of the test to just create an empty cloud name? The behavior is now blocked in the cloud implementation

car-roll commented on src/test/java/hudson/plugins/ec2/AmazonEC2CloudTest.java:
```diff
@@ -92,7 +92,14 @@ public void testSshKeysCredentialsIdRemainsUnchangedAfterUpdatingOtherFields() t
         HtmlTextInput input = form.getInputByName("_.cloudName");
 
         input.setText("test-cloud-2");
-        r.submit(form);
```
For this test, `r.submit(form)` returns 
```
FailingHttpStatusCode 404 Not Found for http://localhost:65397/jenkins/cloud/us-east-1/
```
I'm not sure if this is a regression, but it looks like because we change the name of the cloud node, the url is now dead. If this is acceptable, I did not want to add an `assertThrows` in case changes in the core would make that error go away.

raul-arabaolaza replied:
Sorry, @car-roll I did not fully understood your comment. From what you say it looks like changing the cloud name now is broken? If not what is being done by the production code to deal with that 404 that now is happening? It looks to me that instead of a try catch we should prevent that error to happen in first case

car-roll replied:
Ouch, yeah so I manually tested this and this is a regression. Changing the names in the cloud config page fails on 404 😬

car-roll replied:
yes, with a dedicated rename page

car-roll commented on pom.xml:
```diff
@@ -78,7 +78,7 @@ THE SOFTWARE.
         <revision>2.0.8</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.361.4</jenkins.version>
+        <jenkins.version>2.403</jenkins.version>
```
I chose 2.403 because that was when the changes were first introduced. Most recent LTS is 2.401.3.

jtnord replied:
`2.414` will become the lts line and there is a bom for this line (which you also should update!)

https://github.com/jenkinsci/ec2-plugin/pull/875/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R222-R223

raul-arabaolaza commented on src/test/java/hudson/plugins/ec2/SlaveTemplateTest.java:
```diff
@@ -217,7 +220,8 @@ public void testUnixConfigRoundTrip() throws Exception {
         r.jenkins.clouds.add(ac);
 
-        r.submit(r.createWebClient().goTo("configureClouds").getFormByName("config"));
+        r.submit(getConfigForm(ac));
```
Why submit twice?

car-roll replied:
because i didn't carefully check my find and replace 🤦

car-roll replied:
resolved in ca8172ba72f4cbaa7aadee92e5ad3f336e77034d

car-roll commented:
So `testSshKeysCredentialsIdRemainsUnchangedAfterUpdatingOtherFields` still fails. Looks to be a regression. When you change the cloud node name, the user then receives a 404 saying the URL to the old cloud config page does not exist.
However, the cloud name does change successfully when you go back, it's just that submitting the name change tries to take you back to the old cloud name config page.

car-roll commented:
Turns out the issue is here: https://github.com/jenkinsci/jenkins/blob/31721319231f19bead0adcbb005b99c84a6f9171/core/src/main/java/hudson/slaves/Cloud.java#L334
The redirect is incorrect to return to the top of the clouds page. Will file a PR for fix.

raul-arabaolaza commented:
I beieve this handles properly the UI changes, while the 404 is managed by the core pr, I see no reason not to ship this fixes.

raul-arabaolaza commented on pom.xml:
```diff
@@ -78,7 +78,7 @@ THE SOFTWARE.
         <revision>2.0.8</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.361.4</jenkins.version>
+        <jenkins.version>2.420-rc34091.108b_c22ede78</jenkins.version> <!-- TODO: https://github.com/jenkinsci/jenkins/pull/8310 -->
```
I guess this should be `2.414` to be aligned with the BOM version?

car-roll replied:
sounds good, I will go back to 2.414, but there still will be the failing test. Should I add an '@Ignore` with a comment on the related jenkins ticket then?

raul-arabaolaza replied:
> Should I add an '@Ignore` with a comment on the related jenkins ticket then?

AFAIUI the intent of `testSshKeysCredentialsIdRemainsUnchangedAfterUpdatingOtherFields` is to check the credentials are not changing after editing it, so we could also slightly adjust the test to change other field instead of the cloud name. That way we do not have to ignore the test or block the release. I would add also a comment with the link to the core issue so we restore/add the original test when done.

car-roll replied:
test updated in 3ae03ee064c62110cdefc1d2d209ab325fedc6de

raul-arabaolaza commented on src/test/java/hudson/plugins/ec2/AmazonEC2CloudTest.java:
```diff
@@ -32,6 +32,7 @@
 import com.cloudbees.plugins.credentials.CredentialsStore;
 import com.cloudbees.plugins.credentials.SystemCredentialsProvider;
 import com.cloudbees.plugins.credentials.domains.Domain;
+import org.htmlunit.FailingHttpStatusCodeException;
```
```suggestion
```

Leftover from previous commits now removed I guess

car-roll replied:
removed in 60e1b46a4ae4e2363216adf5e604acfa294512e7

raul-arabaolaza commented on src/test/java/hudson/plugins/ec2/ConfigurationAsCodeTest.java:
```diff
@@ -37,7 +37,7 @@ public class ConfigurationAsCodeTest {
     @ConfiguredWithCode("EC2CloudEmpty.yml")
     public void testEmptyConfig() throws Exception {
-        final AmazonEC2Cloud ec2Cloud = (AmazonEC2Cloud) Jenkins.get().getCloud("");
+        final AmazonEC2Cloud ec2Cloud = (AmazonEC2Cloud) Jenkins.get().getCloud("empty");
```
This is changing the test intent, is it intentional?

car-roll replied:
Actually we had this discussion previously that was resolved :)  https://github.com/jenkinsci/ec2-plugin/pull/875#discussion_r1277677549

The purpose of the test was to create an ec2Cloud with no templates, so using the name "empty" is still okay. There is a check in core now that prevents the creation of a template with no name.

raul-arabaolaza reviewed (commented):
I believe we should try to move this forward

car-roll commented:
@Vlatombe I believe this PR is ready for review now. Would you be able to take a look at it?

car-roll commented:
Also created https://github.com/car-roll/ec2-plugin/pull/1 to handle a separate name change should https://github.com/jenkinsci/jenkins/pull/8310 get in before this PR

car-roll commented:
Sorry forgot to remove an unnecessary TODO (related to name change fixes). I changed the test so it didn't rely on name changes and just changed a different property.

car-roll commented:
dealing with CI flakes

car-roll commented:
@res0nance Could you take a look at this PR? Tests fail here when the jenkins version is upgraded. It also revealed a regression in core (See https://github.com/jenkinsci/jenkins/pull/8310/)

Will continue to work on getting CI to pass. I believe it is infrastructure right now that is making it fail.

car-roll commented:
ci now passes

res0nance commented:
Hi @car-roll changes look fine would you like this merged or do you still have work to do in this area?

car-roll commented:
Hi @res0nance this PR is done so we can merge it 🚀  many thanks!
